salt/modules/win_file.py

```diff
@@ -1584,8 +1584,13 @@ def check_perms(path,
                     ret['comment'].append(
                         'Failed to change owner to "{0}"'.format(owner))
 
-    # Check permissions
-    cur_perms = salt.utils.win_dacl.get_permissions(path)
+    # Permissions will only be set on first call of _get_permissions.
+    # Afterwards the variable from the outer scope will just be retrieved.
+    check_perms.cur_perms = None
+    def _get_permissions():
+        if not check_perms.cur_perms:
+            check_perms.cur_perms = salt.utils.win_dacl.get_permissions(path)
+        return check_perms.cur_perms
```

Comment:
I don't understand the purpose of this code. Could you elaborate on why we need to save the current permissions in an attribute to the check_perms function? I'm not seeing the reason for trying to save perms in between runs. This function needs to check permissions for each path that is passed. It shouldn't be rememberd.

Reply:
If I understand the original code correctly the result of `salt.utils.win_dacl.get_permissions(path)` is *only* used in the following situations (for `check_perms`):
* `if deny_perms is not None:` --> argument for the function
* `if changes` --> this is only set in case `deny_perms` was not None
* `if grant_perms is not None:` --> argument for the function
* `if changes` --> this is only set in case `grant_perms` was not None
*  `if reset` --> argument for the function

In case none of those evaluate to true the return of the function `salt.utils.win_dacl.get_permissions(path)` is not not used. 

This was the case for the original issue - I did not try to change any of the permissions. The runtime still crashed because of a bug inside the evaluation of `get_permissions` (or probably rather the Windows API). 

Therefore my idea was to evaluate `get_permissions` as late as possible. 

The function actually does not save the permission between runs in the code above, since the variable `check_perms.cur_perms` is local to the function `check_perms`. Thus, for every new call of `check_perms` `cur_perms` will initially evaluate to `None` and only be filled on the first call of `_get_permissions()`.

Reply:
OK. The way the state works is that it has to get the current status of the permissions before making any changes. This is to show predicted changes if test=True and to show actual changes after the state run. So it has to get the permissions of the existing object first.

Since permissions are not saved between runs, I see no difference between the original code (`cur_perms = ...`) and the new code. I don't think this change is needed.